### PR TITLE
feat(cache): import external recurrent checkpoints atomically across ranks

### PR DESCRIPTION
**Pull request by voipmonitor**
## Review composition

**Merge #664 and #553 before this PR.** #664 owns connector-readiness polling and prefill-lane release; this PR owns atomic import, publication, cancellation and cache-hit attribution. Both review units are based on `dev/jovian-judgement`; their pinned heads compose without an additional scheduler patch. Do not enable external recurrent checkpoint serving without the scheduler readiness hook.

Status: **implemented; composed CPU contracts qualified**. The JJ-based serving composition passes 398 scheduler, checkpoint-admission, sparse-cleanup, connector and timing tests. This is not an end-to-end qualification of the updated JJ/model dependencies. Original contribution authors remain in the retained commits; integration and validation used OpenAI Codex assistance.

## Behavior

Expose an atomic external-checkpoint API for request-boundary caching. A
connector reserves target attention, recurrent and MTP/DFlash auxiliary pages
privately, copies every rank, then publishes the complete generation. Partial
copies never become a runnable cache hit.

Cancellation keeps destination pages pinned until admitted transfers drain.
Allocation failure releases private reservations without partially publishing
the bundle. A pending restore defers only that request; resident decodes remain
schedulable. Source attribution distinguishes imported tokens from local hits.

The V2 logits-only restore path also dispatches scheduled cache tasks, including
stores belonging to other requests. Otherwise a boundary step can indefinitely
strand their asynchronous transfers.

## Dependencies and compatibility

- #676 supplies authenticated partial-attention reuse; #553 allows the
  validated engine-driven connector with expandable CUDA segments. Both
  prerequisites are included with their original authorship.
- #682 already supplies target/MTP/DFlash boundary state restoration. This
  PR does not duplicate that implementation; it adds the external ownership
  and transfer contract.
- Ordinary aligned connectors retain their separate behavior.
- The standalone branch uses JJ's scheduler. When composed with #664,
  deferring an import must additionally return that scheduler's reserved
  interleave credit; the serving integration retains that operation.

## Validation

Status: **implemented; atomic allocation and transport contracts qualified**.
The branch passes 21 targeted cases against JJ at `f9dc27dde5d`:

```bash
pytest -q tests/v1/core/test_prefix_caching.py tests/v1/core/test_scheduler.py \
  tests/v1/worker/test_gpu_input_batch_v2.py \
  tests/v1/worker/test_gpu_model_runner_v2.py \
  tests/v1/kv_connector/unit/test_config.py \
  -k 'external_boundary_import or invalidated_external_import or boundary_restore_preserves_external or pending_boundary_import or boundary_logits_only_dispatches or external_checkpoint_layout or engine_driven or boundary_hit_is_admitted'
```

Tests cover cancellation, all-rank acknowledgement, pool-capacity rejection,
restart with different physical capacity, imported-token attribution and
admission alongside active decode. Cached OPT configuration metadata supplies
scheduler fixtures; no OPT model weights execute. Initial unavailable-fixture
failures are retained separately. Pre-commit passes.

With the matching LMCache connector, the immutable serving image
`sha256:7af278fa25acd9647943408c1630dac31d35bd57da01044402d8fced6a0b048b`
passes all six no-spec/MTP3/DFlash2 × DCP1/DCP4 FP8 transport arms on stock
RTX PRO 6000 Workstation quartets. Each includes a million-token cold/RAM/
restart restore, exact 54K document answers, shared SYSTEM/different USER
reuse, C4 all-rank byte checks and C8 cancellation/live-lease eviction.
Every million-token restore reports one million external tokens and no local
recomputation. The sidecar creates no CUDA context.

This proves transfer ownership and source attribution, not long-generation
acceptance parity. Balanced speculative acceptance and full-image performance
remain separate release gates. OpenAI Codex assisted under Martin Vít's
direction; human review is requested.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
Review composition validation: merge #664 and #553 before this PR. Readiness polling/lane release are reviewed in #664, not a hidden integration patch. Final composed JJ passes 393 focused CPU tests. The public 32-head vLLM manifest reproduces the complete tested tree with no resolution patch. The packaged serving smoke uses GPU-local cache; fresh external-cache restart qualification is not claimed by it.

Evidence and limitations: https://github.com/local-inference-lab/blackwell-llm-docker/blob/codex/glm53-source-locked-build/recipes/glm53/review-qualification.md

Validation performed with OpenAI Codex assistance.

@coderabbitai review

**Comment by voipmonitor** [comment]
Included in `dev/jovian-judgement` through [this PR's individual merge](https://github.com/local-inference-lab/vllm/commit/5cf75f662c0a19c0670c3e8491becd38dd55a082). The reviewed head and its contributor commits remain ancestors; the PR is merged and closed.

Source validation: replaying all 32 R35 review heads on the pinned base exactly reproduces the released Docker's vLLM tree; all 6,870 installed tracked files match. JJ additionally preserves Luke's DS4.1 work and #734. The final composition passed 247 focused checkpoint/scheduler, sampler/warmup and native GPU tests. This is combined-source evidence, not a fresh performance or full-model qualification for this individual PR.

Publication-history clarification: the individual merge linked above is in JJ's first-parent history. It replaces the receipt's archived wrapper-merge reference; GitHub's historical merge SHA may still identify that archive. See #731 for component review order and qualification limits.
